Velero disaster recovery (Closes #331, Closes #332) (#698)
SUSE/doc-caasp: adoc/admin-disaster-recovery-velero.adoc

= Disaster Recovery

> https://velero.io/[Velero] is a solution for supporting {kube} cluster disaster recovery, data migration and data protection by backing up {kube} cluster resources and persistent volumes to external supported storage backend on demand or by schedule. The major functions include:

Reviewer: ```suggestion
link:https://velero.io/[Velero] is a solution for supporting {kube} cluster disaster recovery, data migration and data protection by backing up {kube} cluster resources and persistent volumes to external supported storage backend on demand or by schedule. The major functions include:
```

* Backup K8s resources and persistent volumes for supported storage providers
* Restore K8s resources and persistent volumes for supported storage providers

> * When backing up persistent volumes w/o supported storage provider, Velero leverages https://github.com/restic/restic[restic] as an agnostic solution to backup this sort of persistent volumes under some known limitations.

Reviewer: ```suggestion
* When backing up persistent volumes w/o supported storage provider, Velero leverages link:https://github.com/restic/restic[restic] as an agnostic solution to backup this sort of persistent volumes under some known limitations.
```

* Backup selected {kube} resources then restore if the selected {kube} resources loss
* Backup selected {kube} resources and persistent volumes then restore if the {kube} selected {kube} resources loss or data loss
* Migrate the backup to another clusters for any purpose like migrates production cluster to develop cluster for testing.

Velero consists of below components:

* A velero server that runs on your {kube} cluster
* A restic deployed on each worker nodes  that runs on your {kube} cluster (optional)
* A command-line client that runs locally

== Prerequisites

. Helm
+
In order to successfully use Velero to backup and restore the {kube} cluster, you first need to install Helm and Tiller. Refer to <<helm_tiller_install>>.

. SUSE Helm Chart
+
Add SUSE helm chart repository URL
+
[source,bash]
----
helm repo add suse https://kubernetes-charts.suse.com
----

. Object Storage
+
Velero uses object storage to store backups and associated artifacts. It can optionally snapshot persistent volume and store in object storage by `restic` if there is no supported volume snapshot provider.

.. Storage Providers
+
|=======
| Provider | Object Storage | Plugin Provider Repo

> | Amazon Web Services (AWS) | AWS S3 | https://github.com/vmware-tanzu/velero-plugin-for-aws[Velero plugin for AWS]

Reviewer: ```suggestion
| Amazon Web Services (AWS) | AWS S3 | link:https://github.com/vmware-tanzu/velero-plugin-for-aws[Velero plugin for AWS]
```

> | Google Cloud Platform (GCP) | Google Cloud Storage | https://github.com/vmware-tanzu/velero-plugin-for-gcp[Velero plugin for GCP]

Reviewer: ```suggestion
| Google Cloud Platform (GCP) | Google Cloud Storage | link:https://github.com/vmware-tanzu/velero-plugin-for-gcp[Velero plugin for GCP]
```

> | Microsoft Azure | Azure Blob Storage | https://github.com/vmware-tanzu/velero-plugin-for-microsoft-azure[Velero plugin for Microsoft Azure]

Reviewer: ```suggestion
| Microsoft Azure | Azure Blob Storage | link:https://github.com/vmware-tanzu/velero-plugin-for-microsoft-azure[Velero plugin for Microsoft Azure]
```

|=======

.. S3-compatible Storage Providers

... SES6 Ceph Object Gateway (radosgw)
+

> SUSE supports the SES6 Ceph Object Gateway (radosgw) as a S3-compatible object storage provider. Refer to https://documentation.suse.com/ses/6/html/ses-all/cha-ceph-additional-software-installation.html[SES6 Object Gateway Manual Installation] on how to install it.

Reviewer: ```suggestion
SUSE supports the SES6 Ceph Object Gateway (radosgw) as a S3-compatible object storage provider. Refer to link:https://documentation.suse.com/ses/6/html/ses-all/cha-ceph-additional-software-installation.html[SES6 Object Gateway Manual Installation] on how to install it.
```

+
... Minio
+

> Besides SUSE enterprise storage, there is an alternative open source S3-compatible object storage provider https://min.io/[minio].  Prepare an external host and install Minio on the host.

Reviewer: ```suggestion
Besides SUSE enterprise storage, there is an alternative open source S3-compatible object storage provider link:https://min.io/[minio].  Prepare an external host and install Minio on the host.
```

+
[source,bash]
----
# Download Minio server
wget https://dl.min.io/server/minio/release/linux-amd64/minio
chmod +x minio

# Expose Minio access_key and secret_key
export MINIO_ACCESS_KEY=<access_key>
export MINIO_SECRET_KEY=<secret_key>

# Start Minio server
mkdir -p bucket
./minio server bucket &

# Download Minio client
wget https://dl.min.io/client/mc/release/linux-amd64/mc
chmod +x mc

# Setup Minio server
./mc config host add velero http://localhost:9000 $MINIO_ACCESS_KEY $MINIO_SECRET_KEY

# Create bucket on Minio server
./mc mb -p velero/velero
----

+

> For the rests of the S3-compatible storage providers, refer to https://velero.io/docs/v1.2.0/supported-providers/[Velero Supported Providers].

Reviewer: ```suggestion
For the rests of the S3-compatible storage providers, refer to link:https://velero.io/docs/v1.2.0/supported-providers/[Velero Supported Providers].
```

. Object Storage Credential

.. Public Cloud Providers

... AWS S3
+
Create the credential file `credentials-velero` in the local machine.
+
```
[default]
aws_access_key_id=<AWS_ACCESS_KEY_ID>
aws_secret_access_key=<AWS_SECRET_ACCESS_KEY>
```
+

> Please refer to https://github.com/vmware-tanzu/velero-plugin-for-aws/tree/v1.0.0[Velero Plugin For AWS].

Reviewer: ```suggestion
Please refer to link:https://github.com/vmware-tanzu/velero-plugin-for-aws/tree/v1.0.0[Velero Plugin For AWS].
```

... Google Cloud Storage
+
Create the credential file `credentials-velero` in the local machine.
+

> Please refer to https://github.com/vmware-tanzu/velero-plugin-for-gcp/tree/v1.0.0[Velero Plugin For GCP].

Reviewer: ```suggestion
Please refer to link:https://github.com/vmware-tanzu/velero-plugin-for-gcp/tree/v1.0.0[Velero Plugin For GCP].
```

... Azure Blob Storage
+
Create the credential file `credentials-velero` in the local machine.
+

> Please refer to https://github.com/vmware-tanzu/velero-plugin-for-microsoft-azure/tree/v1.0.0[Velero Plugin For Azure].

Reviewer: ```suggestion
Please refer to link:https://github.com/vmware-tanzu/velero-plugin-for-microsoft-azure/tree/v1.0.0[Velero Plugin For Azure].
```

.. S3-compatible Storage Providers, like radosgw
+
Create the credential file `credentials-velero` in the local machine.
+
```
[default]
aws_access_key_id=<S3_COMPATIBLE_STORAGE_ACCESS_KEY_ID>
aws_secret_access_key=<S3_COMPATIBLE_STORAGE_SECRET_ACCESS_KEY>
```

. Volume Snapshotter
+
[NOTE]
A Volume snapshotter is able to snapshot it's persistent volumes if it's volume driver supports do volume snapshot. If a volume provider does not support snapshot or does not have supported Velero storage plugin, Velero will leverage _restic_ to do persistent volume backup and restore.
+
[IMPORTANT]
Please be well-known what persistent volumes are supported for snapshot by storage providers. Then administrators have to determine whether to deploy the _restic_ daemonset or not.
+
|=======
| Provider | Volume Snapshotter | Plugin Provider Repo

> | Amazon Web Services (AWS) | AWS EBS | https://github.com/vmware-tanzu/velero-plugin-for-aws[Velero plugin for AWS]

Reviewer: ```suggestion
| Amazon Web Services (AWS) | AWS EBS | link:https://github.com/vmware-tanzu/velero-plugin-for-aws[Velero plugin for AWS]
```

|=======

> For the rests snapshotter providers, refer to https://velero.io/docs/v1.2.0/supported-providers/[Velero Supported Providers]

Reviewer: ```suggestion
For the rests snapshotter providers, refer to link:https://velero.io/docs/v1.2.0/supported-providers/[Velero Supported Providers]
```

* Backup whole {kube} cluster resources then restore if any {kube} resources loss
* Backup selected {kube} resources then restore if the selected {kube} resources loss
* Backup selected {kube} resources and persistent volumes then restore if the {kube} selected {kube} resources loss or data loss
* Migrate the backup to another clusters for any purpose like migrates production cluster to develop cluster for testing.

Velero consists of below components:

* A velero server that runs on your {kube} cluster
* A restic deployed on each worker nodes  that runs on your {kube} cluster (optional)
* A command-line client that runs locally

== Prerequisites

. Helm
+
In order to successfully use Velero to backup and restore the {kube} cluster, you first need to install Helm and Tiller. Refer to <<helm_tiller_install>>.

. SUSE Helm Chart
+
Add SUSE helm chart repository URL
+
[source,bash]
----
helm repo add suse https://kubernetes-charts.suse.com
----

. Object Storage
+
Velero uses object storage to store backups and associated artifacts. It can optionally snapshot the persistent volume and stores in the object storage if the persistent volumes does not supports volume snapshot.

> .. Public Cloud Storage Providers

Author: Updated! Please help on review :bowing_man:

+
SUSE supports the SES6 Ceph Object Gateway (radosgw) as the S3-compatible storage provider. Refer to https://documentation.suse.com/ses/6/html/ses-all/cha-ceph-additional-software-installation.html[SES6 Object Gateway Manual Installation] on how to install it.
+
... Minio
+
For the evaluation install, there are opensource S3-compatible storage provider https://min.io/[minio]. Prepare an external host and install Minio on the host.
+
[source,bash]
----
# Download Minio server
wget https://dl.min.io/server/minio/release/linux-amd64/minio
chmod +x minio

# Expose Minio access_key and secret_key
export MINIO_ACCESS_KEY=<access_key>
export MINIO_SECRET_KEY=<secret_key>

# Start Minio server
mkdir -p bucket
./minio server bucket &

# Download Minio client
wget https://dl.min.io/client/mc/release/linux-amd64/mc
chmod +x mc

# Setup Minio server
./mc config host add velero http://localhost:9000 $MINIO_ACCESS_KEY $MINIO_SECRET_KEY

# Create bucket on Minio server
./mc mb -p velero/velero
----

+
For the rests of the S3-compatible storage providers, refer to https://velero.io/docs/v1.2.0/supported-providers/[Velero Supported Providers].

. Volume Snapshotter
+
[NOTE]

> The volume snapshotter snapshots the persistent volume if the volume supports the snapshot API to do volume snapshot. If the volume does not supports snapshot API or does not have supported providers, Velero leverages _restic_ to do persistent volume backup and restore.

Reviewer 2: ```suggestion
A Volume snapshotter is able to snapshot it's persistent volumes if it's volume driver supports do volume snapshot. If a volume provider does not support snapshot or does not have supported Velero storage plugin, Velero will leverage _restic_ to do persistent volume backup and restore.
```

[IMPORTANT]

> Please be well-known the persistent volume in the supported volume snapshotter provides. Then adminstrators have to determine whether to deploy the _restic daemonset_ or not.

Reviewer 2: ```suggestion
Please be well-known what persistent volumes are supported for snapshot by storage providers. Then administrators have to determine whether to deploy the _restic daemonset_ or not.
```

> .. Public Cloud Storage Providers

Reviewer 2: The same as above suggestion, please add a link to the Velero documentation. https://velero.io/docs/v1.2.0/supported-providers/.

+
|=======
| Provider | Volume Snapshotter | Plugin Provider Repo
| Amazon Web Services (AWS) | AWS EBS | https://github.com/vmware-tanzu/velero-plugin-for-aws[Velero plugin for AWS]
|=======

> .. Community Supported Providers

Reviewer 2: I would suggest no need to list non-supported items and instead just have a link to external documentation.

| OpenEBS | OpenEBS CStor Volume | https://github.com/openebs/velero-plugin[OpenEBS]
| AlibabaCloud | AlibabaCloud | https://github.com/AliyunContainerService/velero-plugin[AlibabaCloud]
| Hewlett Packard | HPE Storage | https://github.com/hpe-storage/velero-plugin[Hewlett Packard	Slack]
|=======

. Object Storage Credential

.. Public Cloud Providers
... AWS S3
+
Create the credential file `credentials-velero` in the local machine.
+
```
[default]
aws_access_key_id=<AWS_ACCESS_KEY_ID>
aws_secret_access_key=<AWS_SECRET_ACCESS_KEY>
```
+
Please refer to https://github.com/vmware-tanzu/velero-plugin-for-aws/tree/v1.0.0[Velero Plugin For AWS].

... Google Cloud Storage
+
Create the credential file `credentials-velero` in the local machine.
+
Please refer to https://github.com/vmware-tanzu/velero-plugin-for-gcp/tree/v1.0.0[Velero Plugin For GCP].

... Azure Blob Storage
+
Create the credential file `credentials-velero` in the local machine.
+
Please refer to https://github.com/vmware-tanzu/velero-plugin-for-microsoft-azure/tree/v1.0.0[Velero Plugin For Azure].

> .. S3-compatible Storage Providers

Reviewer 2: ```suggestion
.. S3-compatible Storage Providers, like Minio
```

+
Create the credential file `credentials-velero` in the local machine.
+
```
[default]
aws_access_key_id=<S3_COMPATIBLE_STORAGE_ACCESS_KEY_ID>
aws_secret_access_key=<S3_COMPATIBLE_STORAGE_SECRET_ACCESS_KEY>
```
+

. Velero CLI
+
[source,bash]
----
sudo zypper install velero
----

> == Limitations

Reviewer 2: Also have a reference link to the Velero documentation for the detailed known issues and limitations.

    --set configuration.backupStorageLocation.name=<BUCKET_NAME> \
    --set configuration.backupStorageLocation.config.region=<REGION_NAME> \
    --set snapshotsEnabled=false \
    --set initContainers[0].name=velero-plugin-for-aws \
    --set initContainers[0].image=registry.suse.com/caasp/v4/velero-plugin-for-aws:1.0.0 \
    --set initContainers[0].volumeMounts[0].mountPath=/target \
    --set initContainers[0].volumeMounts[0].name=plugins \
    suse/velero
----

+
Then, suggests to create at least one additional backup locations point to different object storage server to prevent object storage server single point of failure.

+
[source,bash]
----
velero backup-location create slave \
    --provider aws \
    --bucket <BUCKET_NAME> \
    --config region=<REGION_NAME>
----

* The backup location on S3-compatible storage providers
+
. The backup bucket name _BUCKET_NAME_. (The bucket name in S3 object storage)
. The backup region name _REGION_NAME_. (The region name for the S3 object storage. For example, radosgw _or_ master/slave if you have HA S3 object storage backups)
. The S3-compatible object storage simulates the S3 object storage.

> Therefore, the configuration for S3-compatible object storage have to setup additional configurations

Reviewer 2: ```suggestion
Therefore, the configuration for S3-compatible object storage have to setup additional configurations.
```

. The Velero installed namespace _NAMESPACE_, the default namespace is `velero`. (optional)
+
[NOTE]
The public cloud provider supports persistent volume snapshot API. Therefore, we _do not have to_ deploy the restic daemonset.

+
[source,bash]
----
helm install \
    --name velero \
    --namespace <NAMESPACE> \
    --set-file credentials.secretContents.cloud=credentials-velero \
    --set configuration.provider=aws \
    --set configuration.backupStorageLocation.name=default \
    --set configuration.backupStorageLocation.name=<BUCKET_NAME> \
    --set configuration.backupStorageLocation.config.region=<REGION_NAME> \
    --set snapshotsEnabled=true \
    --set configuration.volumeSnapshotLocation.name=default \
    --set configuration.volumeSnapshotLocation.config.region=<REGION_NAME> \
    --set initContainers[0].name=velero-plugin-for-aws \
    --set initContainers[0].image=registry.suse.com/caasp/v4/velero-plugin-for-aws:1.0.0 \
    --set initContainers[0].volumeMounts[0].mountPath=/target \
    --set initContainers[0].volumeMounts[0].name=plugins \
    suse/velero
----

+

> Then, suggests to create at least one additional backup locations point to different object storage server to prevent object storage server single point of failure.

Reviewer 2: ```suggestion
Then, suggest to create at least one additional backup locations point to different object storage server to prevent object storage server single point of failure.
```

+
[source,bash]
----
helm install \
    --name velero \
    --namespace <NAMESPACE> \
    --set-file credentials.secretContents.cloud=credentials-velero \
    --set configuration.provider=aws \
    --set configuration.backupStorageLocation.name=default \
    --set configuration.backupStorageLocation.name=<BUCKET_NAME> \
    --set configuration.backupStorageLocation.config.region=<REGION_NAME> \
    --set configuration.backupStorageLocation.config.s3ForcePathStyle=true \
    --set configuration.backupStorageLocation.config.s3Url=<S3_COMPATIBLE_STORAGE_SERVER_URL> \
    --set snapshotsEnabled=true \
    --set deployRestic=true \
    --set configuration.volumeSnapshotLocation.name=default \
    --set configuration.volumeSnapshotLocation.config.region=minio \
    --set initContainers[0].name=velero-plugin-for-aws \
    --set initContainers[0].image=registry.suse.com/caasp/v4/velero-plugin-for-aws:1.0.0 \
    --set initContainers[0].volumeMounts[0].mountPath=/target \
    --set initContainers[0].volumeMounts[0].name=plugins \
    suse/velero
----

+

> Then, suggests to create at least one additional backup locations point to different object storage server to prevent object storage server single point of failure.

Reviewer 2: ```suggestion
Then, suggest to create at least one additional backup locations point to different object storage server to prevent object storage server single point of failure.
```

+
However, if we deploy the restic daemonset and want to backup the persistent volume by restic, we have to add annotation `backup.velero.io/backup-volumes=<VOLUME_NAME_1>,<VOLUME_NAME_2>,...` to the pods which have mounted the volume manually.
+
For example, we deploy a Elasticsearch cluster and want to backup the Elasticsearch cluster's data. Add the annotation to the Elasticsearch cluster pods:
+
[source,bash]
----
kubectl annotate pod/elasticsearch-master-0 backup.velero.io/backup-volumes=elasticsearch-master
kubectl annotate pod/elasticsearch-master-1 backup.velero.io/backup-volumes=elasticsearch-master
kubectl annotate pod/elasticsearch-master-2 backup.velero.io/backup-volumes=elasticsearch-master
----
[NOTE]
Velero currently does not provide a mechanism to detect persistent volume claims that are missing the restic backup annotation. To solve this, there is a community provided controller https://github.com/bitsbeats/velero-pvc-watcher[velero-pvc-watcher] which integrate Prometheus to generate alerts for volumes that are not in the backup or backup-exclusion annotation.

** Manual
+
[source,bash]
----
velero backup create <BACKUP_NAME>
----

** Schedule
+

> The schedule templatein in cron notation, using UTC time:

Reviewer 2: ```suggestion
The schedule template in cron notation, using UTC time:
```

# Create a backup excluding the kube-system and default namespaces
velero backup create backup-1 --exclude-namespaces kube-system,default
----

**** Resources
+
Pass flag `--include-resources` or `--exclude-resources` to specifies which resources to include/exclude when backing up. For example:
+
[source,bash]
----
# Create a backup including storageclass resource only
velero backup create backup-1 --include-resources storageclasses
----
+
[TIP]
Use `kubectl api-resources` to lists all API resources on the server.

**** Label Selector
+
Pass `--selector` to only back up resources matching the label selector.
+
[source,bash]
----
# Create a backup for the elasticsearch cluster only
velero backup create backup-1 --selector app=elasticsearch-master
----

*** Location
+
Pass `--storage-location` to specify where the backup stores to. For example, if we have HA object storage server called master and slave respectively.
+
[source,bash]
----
# Create a backup to master storage server
velero backup create backup2master --storage-location master

# Create a backup to slave storage server
velero backup create backup2slave --storage-location slave
----

*** Garbage collection
+

> Pass `--ttl` to determine how long keeps the backup, after that, the backup will be garbage collected. The default backup existed time is 720 hours (30 days).

Reviewer 2: @c3y1huang this is kind of file retention.  (This is just a comment :) )

Reviewer 3: For etcd backup cronjob can control retention at job level but the not the actual etcd snapshot stored. I will add in doc as HINT for user. Ref: https://github.com/SUSE/doc-caasp/pull/682.

velero restore create --from-backup backup-1 --exclude-namespaces kube-system,default
----

**** Resources
+
Pass flag `--include-resources` or `--exclude-resources` to `velero restore create` to specifies which resources to include/exclude when restoring. For example:
+
[source,bash]
----
# create a restore for only persistentvolumeclaims and persistentvolumes within a backup
velero restore create --from-backup backup-1 --include-resources persistentvolumeclaims,persistentvolumes
----
+
[TIP]
Use `kubectl api-resources` to lists all API resources on the server.

**** Label Selector
+
Pass `--selector` to only restore the resources matching the label selector. For example:
+
[source,bash]
----
# create a restore for only the elasticsearch cluster within a backup
velero restore create --from-backup backup-1 --selector app=elasticsearch-master
----

*** Troubleshooting
**** Get restores
+
[source,bash]
----
velero restore get
----
**** Describe restores
+
[source,bash]
----
velero restore describe <RESTORE_NAME_1> <RESTORE_NAME_2> <RESTORE_NAME_3>
----
**** Get restore logs
+
[source,bash]
----
velero restore logs <RESTORE_NAME>
----

== Use Cases

=== Disaster Recovery

> Use schedule backup when backing up and change to backup storage location to read-only mode when restoring.

Reviewer 2: ```suggestion
Use schedule backup for periodical backup. When restoring, need to change backup storage location to read-only mode to avoid incorrect data integrity.
```

> Run the backup periodically and when the {kube} cluster ran into unexpected state, recovered from the backup file.

Reviewer 2: ```suggestion
Run the backup periodically. When the {kube} cluster runs into unexpected state, recover from the backup file.
```
